metrics: fix windowed histogram merging approach

ericharmeling opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/103814.
Fixes https://github.com/cockroachdb/cockroach/issues/98266.

This commit updates the windowed histogram merging approach to add the previous window's histogram bucket counts and sample count to those of the current one. As a result of this change, the histograms will no longer report under-sampled quantile values, and timeseries metrics-derived charts (e.g., the quantile-based SQL service latency charts on the DB console's Metrics page) will more accurately display metrics.

Release note (bug fix): Updated the histogram window merge calculation to more accurately interpolate quantile values. This change will result in smoother, more accurate Metrics charts on the DB Console.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104088)
<!-- Reviewable:end -->

ericharmeling commented:
Here's a screenshot of the DB Console custom metrics and a prometheus metrics query for the same quantile:

![image](https://github.com/cockroachdb/cockroach/assets/27286675/ad7e976a-2a18-4eaa-8f44-b42e713abd1e)

aadityasondhi reviewed (changes requested):
Thanks for doing this! The core logic looks good, just need to extend it to all implementations of the windowed histogram implementation. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NWhppoF47mZrjIpRC1k:b-eagf6o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

___
*[`-- commits` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWhlpvjD63z7h_Qha77):*
Please add the issues this fixes:
- https://github.com/cockroachdb/cockroach/issues/103814
- https://github.com/cockroachdb/cockroach/issues/98266
___
*[`pkg/util/metric/metric.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWhm8dG19O2GVUDJ2cG:-NWhm8dG19O2GVUDJ2cH:b-agmw5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/13f9145b5e1eaa9442a5d7b730e73aabeffb7ee5/pkg/util/metric/metric.go#L388)):*
> ```Go
> // Merging the sample count is necessary to correctly calculate the value at
> // a given quantile with the merged bucket counts.
> func (h *Histogram) ToPrometheusMetricWindowedMerged() *prometheusgo.Metric {
> ```

This should be implemented for all implementations of windowed histograms. Consider making it part of the WindowedHistogram Interface. 
___
*[`pkg/util/metric/metric.go` line 453 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWhmFOr9Psz2ISAi9NU:-NWhmFOr9Psz2ISAi9NV:b77ewe6) ([raw file](https://github.com/cockroachdb/cockroach/blob/13f9145b5e1eaa9442a5d7b730e73aabeffb7ee5/pkg/util/metric/metric.go#L453)):*
> ```Go
> //  2. Since the prometheus client library ensures buckets are in a strictly
> //     increasing order at creation, we do not sort them.
> func (h *Histogram) ValueAtQuantileWindowed(q float64) float64 {
> ```

Similarly, every implementation of this function should be updated for every implementation of windowed histogram. 


<!-- Sent from Reviewable.io -->

Commit:
metrics: fix windowed histogram merging approach

Fixes #103814.
Fixes #98266.

This commit updates the windowed histogram merging approach
to add the previous window's histogram bucket counts and sample
count to those of the current one. As a result of this change,
the histograms will no longer report under-sampled quantile values,
and timeseries metrics-derived charts (e.g., the quantile-based
SQL service latency charts on the DB console's Metrics page) will
more accurately display metrics.

Release note (bug fix): Updated the histogram window merge calculation
to more accurately interpolate quantile values. This change will result
in smoother, more accurate Metrics charts on the DB Console.

Co-authored-by: Aaditya Sondhi <20070511+aadityasondhi@users.noreply.github.com>

ericharmeling reviewed (commented):
TFTR!

I've updated the PR following our discussion offline. I'm a little unsure about the `ManualWindowHistogram` merging implementation. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NWrFMg_2DaKX_6-VMV9:b-k2kkzf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`-- commits` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWhlpvjD63z7h_Qha77):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Please add the issues this fixes:
- https://github.com/cockroachdb/cockroach/issues/103814
- https://github.com/cockroachdb/cockroach/issues/98266
</blockquote></details>

Done.
___
*[`pkg/util/metric/metric.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWhm8dG19O2GVUDJ2cG:-NWhv0qkBri5tCBOcu-q:b-zd5st) ([raw file](https://github.com/cockroachdb/cockroach/blob/13f9145b5e1eaa9442a5d7b730e73aabeffb7ee5/pkg/util/metric/metric.go#L388)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

This should be implemented for all implementations of windowed histograms. Consider making it part of the WindowedHistogram Interface. 
</blockquote></details>

After our discussion offline, I believe that we decided to not add `ToPrometheusMetricWindowed`/`ToPrometheusMetricWindowedMerged` to the `WindowedHistogram` interface, and instead rely on `ValueAtQuantileWindowed` implementations to merge prev and cur windows. I've added a note in the interface definition.
___
*[`pkg/util/metric/metric.go` line 453 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWhmFOr9Psz2ISAi9NU:-NWi7d8A8kdfWPwvYzox:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13f9145b5e1eaa9442a5d7b730e73aabeffb7ee5/pkg/util/metric/metric.go#L453)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Similarly, every implementation of this function should be updated for every implementation of windowed histogram. 
</blockquote></details>

Done.
___
*[`pkg/util/metric/metric.go` line 609 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWrGnRP0e2y4iR8LTxz:-NWrGnRP0e2y4iR8LTy-:bbnjqus) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7b0f4785580e27b0e79836b0b52bfd8c9c73b34/pkg/util/metric/metric.go#L609)):*
> ```Go
> func (mwh *ManualWindowHistogram) ToPrometheusMetricWindowed() *prometheusgo.
> 	Metric {
> 	cur := &prometheusgo.Metric{}
> ```

@aadityasondhi 
Does this look right to you?


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (changes requested):
Overall looks great! Just a few minor changes are needed here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NWsDBAy0jnFbIbbSweh:booti4a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

___
*[`pkg/util/metric/metric.go` line 609 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWrGnRP0e2y4iR8LTxz:-NWs8WQB8JZxHH7PUioB:bh943eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7b0f4785580e27b0e79836b0b52bfd8c9c73b34/pkg/util/metric/metric.go#L609)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

@aadityasondhi 
Does this look right to you?
</blockquote></details>

Yes, looks good!
___
*[`pkg/util/metric/metric.go` line 274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsCkbOBbI6Guac0Ibb:-NWsCkbOBbI6Guac0Ibc:bshofmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric.go#L274)):*
> ```Go
> 	h.windowed.tickHelper = &tickHelper{
> 		nextT:        now(),
> 		tickInterval: windowDuration,
> ```

You want to divide the `windowDuration` by 2 since we have 2 windows here. This is similar to what happens in `newHdrHistogram` with the `histWrapNum`. 
___
*[`pkg/util/metric/metric_test.go` line 337 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsBPec9Y7sI2mmrjg8:-NWsBPed9b535CubdT4k:b-13ju2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric_test.go#L337)):*
> ```Go
> 		}
> 		// Tick. This rotates the histogram.
> 		setNow(time.Duration(i+1) * 10 * time.Second)
> ```

I don't think you want 10 seconds here if you have set `Duration` to `time.minute`. You would want to set time to the next `tickInterval + 1`. 
___
*[`pkg/util/metric/metric_test.go` line 341 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsAQAMC1ef_c_D3YfO:-NWsAQAMC1ef_c_D3YfP:b-nkwvib) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric_test.go#L341)):*
> ```Go
> 	}
> 
> 	act := *h.ToPrometheusMetric().Histogram
> ```

Don't you want to call the `ToPrometheusMetricWindowed()` function here? Right now, this test actually tests the cumulative histogram working correctly, which is not the point. If you call `ToPrometheusMetricWindowed()`, I think you will get a test that will actually just show the prev+cur histogram merged view rather than the view of all 4 histograms you have created here. The oldest two histogram views would be removed from the windowed histogram (as designed).


<!-- Sent from Reviewable.io -->

ericharmeling reviewed (commented):
TFTR! I've made some changes following your feedback, with some commentary. I might want to pair offline once more to clarify some things.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NWsE-ZVBby2s9grsfhS:b37aj0t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/util/metric/metric.go` line 274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsCkbOBbI6Guac0Ibb:-NWsEawJ62Qns7FCoEm7:b1waxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric.go#L274)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

You want to divide the `windowDuration` by 2 since we have 2 windows here. This is similar to what happens in `newHdrHistogram` with the `histWrapNum`. 
</blockquote></details>

I don't fully understand why this is the case, but done. Perhaps we chat offline a bit?
___
*[`pkg/util/metric/metric_test.go` line 337 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsBPec9Y7sI2mmrjg8:-NWxHUs-Aj5P1-IlUMq4:boloiy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric_test.go#L337)):*
> I don't think you want 10 seconds here if you have set `Duration` to `time.minute`.

I meant to set the `Duration` to 10 seconds. Fixed that. (Although I guess it's arbitrary, since we are manually setting the `now` time)

> You would want to set time to the next `tickInterval + 1`.

I think I want to set this to `time.Duration(i+1) * tickInterval`, right? If we are incrementing the `nextTick` value by the `tickInterval` at each tick call, we should be setting the `now` time to exactly when the `maybeTick` evaluates `nextTick().Before(now())`.
___
*[`pkg/util/metric/metric_test.go` line 341 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsAQAMC1ef_c_D3YfO:-NWsPylT3wp4zeO2NiGF:be12c4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric_test.go#L341)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Don't you want to call the `ToPrometheusMetricWindowed()` function here? Right now, this test actually tests the cumulative histogram working correctly, which is not the point. If you call `ToPrometheusMetricWindowed()`, I think you will get a test that will actually just show the prev+cur histogram merged view rather than the view of all 4 histograms you have created here. The oldest two histogram views would be removed from the windowed histogram (as designed).
</blockquote></details>

Oh you're totally right. We should be looking at the windows and not the cumulative histogram. 

But I don't think we need to call `ToPrometheusMetricWindowed` here. The `ToPrometheusMetric` erroneously used here is an implementation of the `PrometheusExportable` interface, which I don't think we want to include `ToPrometheusMetricWindowed`, as this interface is implemented by metrics other than histograms (right?).

`ValueAtQuantileWindowed` calls `ToPrometheusMetricWindowed()` when we start asserting values. I think the trick here is checking the "windowed" functions before and after observations within each window interval. This way we can make sure that the merge is happening and the quantile values are capturing the previous window early on in the current window's interval (i.e., before any observations) and later on (i.e., after all observations).

Does that make sense?


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (approved):
Thanks for all the work, this looks good to me! 
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NXG5DLWECzKpYNxlLjG:b-cv4zyt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ericharmeling)

___
*[`pkg/util/metric/metric.go` line 274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsCkbOBbI6Guac0Ibb:-NXG5A0WAfkRDE6aCqAH:b-2bckbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric.go#L274)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

I don't fully understand why this is the case, but done. Perhaps we chat offline a bit?
</blockquote></details>

Spoke offline about this,
___
*[`pkg/util/metric/metric_test.go` line 337 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsBPec9Y7sI2mmrjg8:-NXB41pVASGNmtzOW0Mn:bhxfoyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric_test.go#L337)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

> I don't think you want 10 seconds here if you have set `Duration` to `time.minute`.

I meant to set the `Duration` to 10 seconds. Fixed that. (Although I guess it's arbitrary, since we are manually setting the `now` time)

> You would want to set time to the next `tickInterval + 1`.

I think I want to set this to `time.Duration(i+1) * tickInterval`, right? If we are incrementing the `nextTick` value by the `tickInterval` at each tick call, we should be setting the `now` time to exactly when the `maybeTick` evaluates `nextTick().Before(now())`.
</blockquote></details>

Yes that seems right to me.
___
*[`pkg/util/metric/metric_test.go` line 341 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsAQAMC1ef_c_D3YfO:-NXG56l39mTvcsboFYvw:bx270d3) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric_test.go#L341)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

Oh you're totally right. We should be looking at the windows and not the cumulative histogram. 

But I don't think we need to call `ToPrometheusMetricWindowed` here. The `ToPrometheusMetric` erroneously used here is an implementation of the `PrometheusExportable` interface, which I don't think we want to include `ToPrometheusMetricWindowed`, as this interface is implemented by metrics other than histograms (right?).

`ValueAtQuantileWindowed` calls `ToPrometheusMetricWindowed()` when we start asserting values. I think the trick here is checking the "windowed" functions before and after observations within each window interval. This way we can make sure that the merge is happening and the quantile values are capturing the previous window early on in the current window's interval (i.e., before any observations) and later on (i.e., after all observations).

Does that make sense?
</blockquote></details>

Spoke offline, seems right.


<!-- Sent from Reviewable.io -->

aadityasondhi commented:
Note: this should probably be backported to 22.2 and 23.1. But I leave that decision to the obs-inf team

ericharmeling reviewed (commented):
@aadityasondhi Thank you for all of the help on this!

> this should probably be backported to 22.2 and 23.1. But I leave that decision to the obs-inf team

@dhartunian Any objections to backporting?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NXHQzoW9-Ath-Mz7Qgq:b-uybk0u)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi and @dhartunian)



<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
Yes, this should be backported to both since both versions have the new histogram by default.

> Reviewed 1 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NXL53nUAtZgxXNZ4srN:bkvbyvq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi and @ericharmeling)

___
*[`pkg/util/metric/metric.go` line 274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsCkbOBbI6Guac0Ibb:-NXQFhVS8FRsGJcEKjrJ:b4ba7es) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric.go#L274)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Spoke offline about this,
</blockquote></details>

@ericharmeling can you document the explanation in a comment right above this field setting?
___
*[`pkg/util/metric/metric.go` line 315 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTAkFx91bnC0ZQtdS6:-NXTAkFx91bnC0ZQtdS7:b-pvl956) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L315)):*
> ```Go
> 	// but that is now how we have set it up right now. It should be doable
> 	// though, since there is only one consumer of windowed histograms - our
> 	// internal timeseries system.
> ```

Is this comment no longer necessary? It seems like we're implemented some version of this.



_Code quote:_
```Go
	// TODO(obs-inf): the way we implement windowed histograms is not great. If
	// the windowed histogram is pulled right after a tick, it will be mostly
	// empty. We could add a third bucket and represent the merged view of the two
	// most recent buckets to avoid that. Or we could "just" double the rotation
	// interval (so that the histogram really collects for 20s when we expect to
	// persist the contents every 10s). Really it would make more sense to
	// explicitly rotate the histogram atomically with collecting its contents,
	// but that is now how we have set it up right now. It should be doable
	// though, since there is only one consumer of windowed histograms - our
	// internal timeseries system.
```
___
*[`pkg/util/metric/metric.go` line 391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXT8hiM5_usQKTN-Pb0:-NXT8hiM5_usQKTN-Pb1:b-b9rt62) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L391)):*
> ```Go
> 		panic(err)
> 	}
> 	if h.windowed.prev != nil {
> ```

why do we need to check for nil here? it seems like `h.windowed.cur` and `h.windowed.prev` are both set in the `onTick` field of `tickHelper`. Why check one for nil and not the other?
___
*[`pkg/util/metric/metric.go` line 616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTBAjMDJHD_3vAFQod:-NXTBAjMDJHD_3vAFQoe:b-23ioli) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L616)):*
> ```Go
> // ToPrometheusMetricWindowed returns a filled-in prometheus metric of the
> // right type.
> func (mwh *ManualWindowHistogram) ToPrometheusMetricWindowed() *prometheusgo.
> ```

nit: maybe suffix method with `Locked` to clarify that lock should be taken prior.
___
*[`pkg/util/metric/metric.go` line 617 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTGLOMDKEdEDPQTRf5:-NXTGLOMDKEdEDPQTRf6:b-vfxf2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L617)):*
> ```Go
> // right type.
> func (mwh *ManualWindowHistogram) ToPrometheusMetricWindowed() *prometheusgo.
> 	Metric {
> ```

nit: why is `Metric {` on a separate line?
___
*[`pkg/util/metric/metric.go` line 639 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTGfti6AC9wKBnJajZ:-NXTGfti6AC9wKBnJaj_:brv1r19) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L639)):*
> ```Go
> func (mwh *ManualWindowHistogram) Total() (int64, float64) {
> 	mwh.mu.RLock()
> 	defer mwh.mu.RUnlock()
> ```

`ToPrometheusMetric` locks this mutex already
___
*[`pkg/util/metric/metric.go` line 653 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTGoszAA7Jq-Dx47J2:-NXTGoszAA7Jq-Dx47J3:brv1r19) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L653)):*
> ```Go
> func (mwh *ManualWindowHistogram) Mean() float64 {
> 	mwh.mu.RLock()
> 	defer mwh.mu.RUnlock()
> ```

`ToPrometheusMetric` locks this mutex already
___
*[`pkg/util/metric/metric.go` line 927 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTG0O0796EoE6fPY4f:-NXTG0O0796EoE6fPY4g:b3eaneg) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L927)):*
> ```Go
> // histogram.
> // NB: Buckets on each histogram must be the same
> func MergeWindowedHistogram(cur *prometheusgo.Histogram, prev *prometheusgo.Histogram) {
> ```

Can you add some unit tests for this merge function?


<!-- Sent from Reviewable.io -->

ericharmeling reviewed (commented):
TFTR @dhartunian ! I think I got to all of your comments.

> Yes, this should be backported to both since both versions have the new histogram by default.

Added the backport labels.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NXVnWFg3s3d6HEuNPNN:b-2qndyn)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @dhartunian, and @tbg)

___
*[`pkg/util/metric/metric.go` line 274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NWsCkbOBbI6Guac0Ibb:-NXW52blCXHE440II151:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb2872c74dfc86d307ba3b27811605da76d65854/pkg/util/metric/metric.go#L274)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

@ericharmeling can you document the explanation in a comment right above this field setting?
</blockquote></details>

Done.
___
*[`pkg/util/metric/metric.go` line 315 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTAkFx91bnC0ZQtdS6:-NXVcjwWEfxtWBO3iXUO:b-z7lpq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L315)):*
I believe @tbg wrote this in https://github.com/cockroachdb/cockroach/commit/eb82a437f09225de8dcaa4b5f65328eb7cf0b4bd. 

> If the windowed histogram is pulled right after a tick, it will be mostly empty.

This is no longer true. However, the comments after this might still apply for future improvements. 

> We could add a third bucket and represent the merged view of the two most recent buckets to avoid that.

We are *kind of* doing this now. But we're really just calculating this "third bucket" every time we call `TotalWindowed`, `MeanWindowed`, and `ValueAtQuantileWindowed`.

@tbg any objections to removing this comment (or folding it into a tracking issue)?
___
*[`pkg/util/metric/metric.go` line 391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXT8hiM5_usQKTN-Pb0:-NXW6gQu3Up5_eCxYZ-X:b-325tau) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L391)):*
We get a runtime error without this check at cluster start-up:

```
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | runtime.panicmem
*   | 	GOROOT/src/runtime/panic.go:260
*   | runtime.sigpanic
*   | 	GOROOT/src/runtime/signal_unix.go:835
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Histogram).ToPrometheusMetricWindowed
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/metric.go:394
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Histogram).MeanWindowed
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/metric.go:435
*   | github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric.(*AggHistogram).MeanWindowed
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric/histogram.go:76
*   | github.com/cockroachdb/cockroach/pkg/server/status.extractValue
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:617
*   | github.com/cockroachdb/cockroach/pkg/server/status.eachRecordableValue.func1
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:646
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Registry).Each.func1
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/registry.go:171
*   | github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric.(*AggHistogram).Inspect
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric/histogram.go:62
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Registry).Each
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/registry.go:170
*   | github.com/cockroachdb/cockroach/pkg/server/status.eachRecordableValue
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:645
*   | github.com/cockroachdb/cockroach/pkg/server/status.(*MetricsRecorder).GenerateNodeStatus
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:509
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).writeNodeStatus.func1
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1056
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:319
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).writeNodeStatus
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1055
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).startWriteNodeStatus.func1
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1013
*   | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetry.func1
*   | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:121
*   | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetryEx[...]
*   | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:142
*   | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetry
*   | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:120
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).startWriteNodeStatus
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1010
*   | github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart
*   | 	github.com/cockroachdb/cockroach/pkg/server/server.go:1945
*   | github.com/cockroachdb/cockroach/pkg/cli.createAndStartServerAsync.func1.2
*   | 	github.com/cockroachdb/cockroach/pkg/cli/start.go:805
*   | github.com/cockroachdb/cockroach/pkg/cli.createAndStartServerAsync.func1
*   | 	github.com/cockroachdb/cockroach/pkg/cli/start.go:853
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_arm64.s:1172
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
```

Checking the `prev` nil also helps us avoid evaluating the merge when one isn't possible (i.e., when there is no previous window).

> Why check one for nil and not the other?

I believe `h.windowed.cur` starts as `nil`, so [when it's rotated on the first call to `newHistogram`](https://github.com/cockroachdb/cockroach/blob/f7006be0e78decd181039d91d8c9412875043051/pkg/util/metric/metric.go#L287-L292), that `nil` value is assigned to `h.windowed.prev` (and a new histogram is assigned to `h.windowed.prev`).

So... should we leave this check here, or explicitly initialize both `prev` and `cur` (and check the merge on an empty `prev` instead)?
___
*[`pkg/util/metric/metric.go` line 616 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTBAjMDJHD_3vAFQod:-NXVo4964cJxtxRgm9VV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L616)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: maybe suffix method with `Locked` to clarify that lock should be taken prior.
</blockquote></details>

Done.
___
*[`pkg/util/metric/metric.go` line 617 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTGLOMDKEdEDPQTRf5:-NXVdNl0EAVNeePmOmhK:b4dtuay) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L617)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: why is `Metric {` on a separate line?
</blockquote></details>

Linter thing - fixed.
___
*[`pkg/util/metric/metric.go` line 639 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTGfti6AC9wKBnJajZ:-NXVdZNy-37lOSGlO_P6:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L639)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

`ToPrometheusMetric` locks this mutex already
</blockquote></details>

Fixed.
___
*[`pkg/util/metric/metric.go` line 653 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTGoszAA7Jq-Dx47J2:-NXVdb6mEm9mF9b6c9dR:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L653)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

`ToPrometheusMetric` locks this mutex already
</blockquote></details>

Fixed.
___
*[`pkg/util/metric/metric.go` line 927 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXTG0O0796EoE6fPY4f:-NXVztCp6_W0hVQxJl8h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L927)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can you add some unit tests for this merge function?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NXpwNqe4tbvwiSP2fDR:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @ericharmeling, and @tbg)

___
*[`pkg/util/metric/metric.go` line 391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXT8hiM5_usQKTN-Pb0:-NXpvOyfA4SGjfnBE1by:bwxh1nf) ([raw file](https://github.com/cockroachdb/cockroach/blob/13beab0ed4a3bc1349d8b00c1e3fba8a1d015f82/pkg/util/metric/metric.go#L391)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

We get a runtime error without this check at cluster start-up:

```
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | runtime.panicmem
*   | 	GOROOT/src/runtime/panic.go:260
*   | runtime.sigpanic
*   | 	GOROOT/src/runtime/signal_unix.go:835
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Histogram).ToPrometheusMetricWindowed
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/metric.go:394
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Histogram).MeanWindowed
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/metric.go:435
*   | github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric.(*AggHistogram).MeanWindowed
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric/histogram.go:76
*   | github.com/cockroachdb/cockroach/pkg/server/status.extractValue
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:617
*   | github.com/cockroachdb/cockroach/pkg/server/status.eachRecordableValue.func1
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:646
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Registry).Each.func1
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/registry.go:171
*   | github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric.(*AggHistogram).Inspect
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/aggmetric/histogram.go:62
*   | github.com/cockroachdb/cockroach/pkg/util/metric.(*Registry).Each
*   | 	github.com/cockroachdb/cockroach/pkg/util/metric/pkg/util/metric/registry.go:170
*   | github.com/cockroachdb/cockroach/pkg/server/status.eachRecordableValue
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:645
*   | github.com/cockroachdb/cockroach/pkg/server/status.(*MetricsRecorder).GenerateNodeStatus
*   | 	github.com/cockroachdb/cockroach/pkg/server/status/recorder.go:509
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).writeNodeStatus.func1
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1056
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:319
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).writeNodeStatus
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1055
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).startWriteNodeStatus.func1
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1013
*   | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetry.func1
*   | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:121
*   | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetryEx[...]
*   | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:142
*   | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetry
*   | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:120
*   | github.com/cockroachdb/cockroach/pkg/server.(*Node).startWriteNodeStatus
*   | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1010
*   | github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart
*   | 	github.com/cockroachdb/cockroach/pkg/server/server.go:1945
*   | github.com/cockroachdb/cockroach/pkg/cli.createAndStartServerAsync.func1.2
*   | 	github.com/cockroachdb/cockroach/pkg/cli/start.go:805
*   | github.com/cockroachdb/cockroach/pkg/cli.createAndStartServerAsync.func1
*   | 	github.com/cockroachdb/cockroach/pkg/cli/start.go:853
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_arm64.s:1172
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
```

Checking the `prev` nil also helps us avoid evaluating the merge when one isn't possible (i.e., when there is no previous window).

> Why check one for nil and not the other?

I believe `h.windowed.cur` starts as `nil`, so [when it's rotated on the first call to `newHistogram`](https://github.com/cockroachdb/cockroach/blob/f7006be0e78decd181039d91d8c9412875043051/pkg/util/metric/metric.go#L287-L292), that `nil` value is assigned to `h.windowed.prev` (and a new histogram is assigned to `h.windowed.prev`).

So... should we leave this check here, or explicitly initialize both `prev` and `cur` (and check the merge on an empty `prev` instead)?
</blockquote></details>

ah makes sense. thanks for explaining in-depth.
___
*[`pkg/util/metric/metric_test.go` line 437 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXpvrkL1yX38hTI9H7e:-NXpvrkL1yX38hTI9H7f:b-83expp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc28e2a653dd26b90a505263bb64a570c33c190/pkg/util/metric/metric_test.go#L437)):*
> ```Go
> 	// sample count, and per-bucket cumulative count values,
> 	// plus those of the previous histogram.
> 	require.Equal(t, *cur.Histogram.SampleCount, uint64(2))
> ```

nit: keep the order consistent between "expected" and "actual" args in the `require` call.


<!-- Sent from Reviewable.io -->

ericharmeling reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-:-NXq6ejW0G8GjudWizaU:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @aadityasondhi, @dhartunian, and @tbg)

___
*[`pkg/util/metric/metric_test.go` line 437 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/104088#-NXpvrkL1yX38hTI9H7e:-NXq6dXQAwfH7cijGl1C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc28e2a653dd26b90a505263bb64a570c33c190/pkg/util/metric/metric_test.go#L437)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: keep the order consistent between "expected" and "actual" args in the `require` call.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ericharmeling commented:
bors r+
